odl/operator/default_ops.py

```diff
@@ -579,6 +585,211 @@ def __str__(self):
         return "{}.T".format(self.vector)
 
 
+class NormOperator(Operator):
+    """Operator taking the norm of a vector.
+
+    ``NormOperator(space)(x) <==> space.norm(x)``
+
+    This is only applicable in normed spaces.
+
+    See Also
+    --------
+    InnerProductOperator : Inner product with fixed vector.
+    DistOperator : Distance to fixed vector.
+    """
+
+    def __init__(self, space):
+        """Initialize a NormOperator instance.
+
+        Parameters
+        ----------
+        space : `LinearSpace`
+            Space to take the norm in.
+        """
+        super().__init__(space, RealNumbers(), linear=False)
+
+    def _call(self, x):
+        """Return the norm of x.
+
+        Parameters
+        ----------
+        x : `domain` `element`
+            A vector in the domain.
+
+        Returns
+        -------
+        out : ``field`` `element`
+            Result of the inner product calculation
+
+        Examples
+        --------
+        >>> import odl
+        >>> r2 = odl.rn(2)
+        >>> op = NormOperator(r2)
+        >>> op([3, 4])
+        5.0
+        """
+        return x.norm()
+
+    def derivative(self, point):
+        """The derivative operator.
+
+        This is only applicable in inner product spaces.
+
+        Parameters
+        ----------
+        x : `domain` `element-like`
+            Point in which to take the derivative.
+
+        Returns
+        -------
+        derivative : `InnerProductOperator`
+
+        Raises
+        ------
+        ValueError
+            If ``point.norm() == 0``, in which case the derivative is not well
+            defined in the Frechet sense.
+
+        Notes
+        -----
+        The derivative cannot be written in a general sense except in Hilbert
+        spaces, in that case, it is given by
+
+        .. Math::
+
+            (D ||.||)(x)(y) = < x / ||x||, y >
+
+        Examples
+        --------
+        >>> import odl
+        >>> r3 = odl.rn(3)
+        >>> op = NormOperator(r3)
+        >>> derivative = op.derivative([1, 0, 0])
+        >>> derivative([1, 0, 0])
+        1.0
+        """
+        point = self.domain.element(point)
+        norm = point.norm()
+        if norm == 0:
+            raise ValueError('not differentiable in 0')
+
+        return InnerProductOperator(point / norm)
```

Comment:
This assumes quite a bit on the norm of the space. The inner product operator needs to use the same weighting as the norm - I guess that's okay. But a custom norm implementation cannot be handled consistently, that needs to be filtered out (and documented that it won't work).